components/rollup_node/p2p_interface.md

```diff
@@ -0,0 +1,206 @@
+# P2P Interface
+
+This peer-to-peer (P2P) network service in the rollup node is optional,
+but greatly improves the latency between the view of sequencers and the rest of the network,
+as well as enabling faster historical sync to be bootstrapped, without relying on a single endpoint.
+
+In summary, the P2P stack looks like:
+- Discovery to find peers: [Discv5][discv5]
+- Connections, peering, transport security, multi-plexing, gossip: [LibP2P][libp2p]
+
+This document only specifies the composition and configuration of these network libraries.
+These components have their own standards, implementations in Go/Rust/Java/Nim/JS/more,
+and adopted several other blockchains, most notably the [L1 consensus layer (Eth2)][eth2-p2p].
+
+## P2P configuration
+
+### Identification
+
+Nodes have a **separate** network- and consensus-identity.
+The network identity is a `secp256k1` key, used for both discovery and active LibP2P connections.
+
+Common representations of network identity:
+- `PeerID`: a LibP2P specific ID derived from the pubkey (through protobuf encoding, typing and hashing)
+- `NodeID`: a Discv5 specific ID derived from the pubkey (through hashing, used in the DHT)
+- `Multi-address`: an unsigned address, containing: IP, TCP port, PeerID
+- `ENR`: a signed record used for discovery, containing: IP, TCP port, UDP port, signature (pubkey can be derived) and L2 network identification. Generally encoded in base64.
+
+### Discv5
+
+#### Structure
+
+The Ethereum Node Record (ENR) for an Optimism rollup node must contain keyed values:
+
+- An IPv4 address (`ip` field) and/or IPv6 address (`ip6` field).
+- A TCP port (`tcp` field) representing the local libp2p listening port.
+- A UDP port (`udp` field) representing the local discv5 listening port.
+- An Optimism (`optimism` field) L2 network identifier
+
+The `optimism` value is encoded as the concatenation of:
+- chain ID (`varint`)
+- fork ID (`varint`)
```

Comment:
We probably want to explain / link to what those are.

It would also be interestig to know why Optimism needs those.

- chain ID: I don't really know how discv5 works, but I'm assuming these mesages will intermingle with discv5 messages sent by L1 nodes and this value signals L2 capabilities?
     - does the presence of the optimism identifier signals that the node should not receive L1 messages, or will it receive both L1 and L2 messages?
- fork ID: why does L2 need this but not L1?

Reply:
> We probably want to explain / link to what those are.

`varint` is very standard, it's very common in language standard libraries, and part of protobuf (google encoding). I'll reference that. The idea is to save some bytes for something that technically can be 256 bits, but generally is < 1 byte.

> I'm assuming these mesages will intermingle with discv5 messages sent by L1 nodes and this value signals L2 capabilities? 

Exactly. The DHT is more secure the bigger it is (harder to censor searching paths, more capacity), so the discovery layer is designed to be shared between protocols (already in use by both Eth1, Eth2, and the portal network in the future)

L1 has its own versioning system. Eth2 uses Forkdigests (4 bytes of hashed versioning data), and Eth1 uses an ugly combination of chain IDs, network IDs that are supposed to match, and devp2p protocol versions. In the case of this L2 we want to not collide with others (possibly copies of optimism, or just our own testnets) and support breaking changes with future upgrade, without changing the identity of the network, so it will need to contain the chain ID as well as a versioning identifier. We could hash those together like Eth2 does, but it only becomes less readable that way.

Reply:
I'd personally never heard of `varint` before. A quick google search seems to show it's indeed a protobuf concept that has spread a bit.

There is a good definition [here](https://go.dev/src/encoding/binary/varint.go) for anyone reading. I think at minimum we should link to that.

We could also just remove the annotation, since we're simply inheriting another spec and this is an explanation of what it mandates. It would also be more consistent, since we don't specify for instance how the IPs are encoded.